### PR TITLE
RWAHS-309 Search result controls

### PR DESCRIPTION
**Pull request by leftclickben**
+ Add controls to search results: pagination and view mode
+ Some refactors around the composition of search results and detail views
+ Requires https://github.com/rwahs/providence-configuration/pull/24

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -27,6 +27,7 @@
     "jquery": "^2.2.4",
     "lodash": "^4.13.1",
     "knockout": "^3.4.0",
-    "text": "requirejs-text#^2.0.15"
+    "text": "requirejs-text#^2.0.15",
+    "querystring": "^3.46.1"
```
Comment (leftclickben) [inline bower.json]: Getting the individual query parameters from the URL in the browser is surprisingly difficult.

File: public/app/config/components.js [hunk 2]
```diff
@@ -8,6 +8,79 @@
         function (ko) {
             return function () {
                 ko.components.register(
+                    'collections/library/detail',
+                    { template: { require: 'text!ui/components/collections/library/detail.html' } }
```
Comment (leftclickben) [inline public/app/config/components.js]: When a component doesn't specify a `viewModel`, only the `template`, then it is bound to whatever is passed as the `parameters`.

File: public/app/config/routes.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
+(function () {
+    'use strict';
+
+    define(
+        [
+            'page',
+            'querystring',
+            'util/bind'
+        ],
+        function (page, qs, bind) {
```
Comment (leftclickben) [inline public/app/config/routes.js]: This module (`routes`) is the RWAHS-specific routing / URL configuration.  The `bind` module is the generic part that calls `ko.applyBindings` and runs the various hooks.  These two together used to be the `router` module.

File: public/app/config/settings.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+(function () {
+    'use strict';
+
+    define(
+        [
+            'util/container'
+        ],
+        function (container) {
+            return function () {
+                container.register('settings.library', {
```
Comment (leftclickben) [inline public/app/config/settings.js]: These are just better centralised now in their own DI components, one per collection.

File: public/app/models/ListPager.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
+(function () {
+    'use strict';
+
+    define(
+        [
+            'lodash',
+            'knockout'
+        ],
+        function (_, ko) {
```
Comment (leftclickben) [inline public/app/models/ListPager.js]: This module takes a big list and presents a smaller list that represents a page at a time of the big list.

File: public/app/models/ListPager.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            var PAGINATION_JUMP_COUNT = 7,
```
Comment (leftclickben) [inline public/app/models/ListPager.js]: Total number of "buttons" to display to numbered pages in the pagination.

File: public/app/models/ListPager.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+                this.pageCount = ko.pureComputed(function () {
+                    return Math.ceil(sourceObservable().length / this.pageSize());
+                }.bind(this));
+
+                this.pageNumber = ko.pureComputed(function () {
+                    return Math.floor(this.start() / this.pageSize());
+                }.bind(this));
+
+                this.currentPage = ko.pureComputed(function () {
```
Comment (leftclickben) [inline public/app/models/ListPager.js]: This is the observable that returns the paged data.

File: public/app/ui/components/collections/library/detail.html [hunk 8]
```diff
@@ -1,19 +1,7 @@
...
-        Loading Library record, please wait...
-    </div>
-    <!-- /ko -->
-    <!-- ko if: displayRecord -->
-    <div class="type-header">Library record</div>
-    <h1>
-        <span data-bind="component: displayFor('Title')"></span>
-        <span class="idno" data-bind="text: idno"></span>
-    </h1>
+<section class="detail detail-library">
```
Comment (leftclickben) [inline public/app/ui/components/collections/library/detail.html]: These templates have been split.  There is `app/ui/pages/detail/index.html` which is generic, and just has the header and then includes the relevant per-collection one of these, that are now at `app/ui/components/collections/[collection]/detail.html`.

File: public/app/ui/components/collections/library/list-result.html [hunk 9]
```diff
@@ -0,0 +1,4 @@
+<div class="list-result list-result-library">
```
Comment (leftclickben) [inline public/app/ui/components/collections/library/list-result.html]: This is the per-collection view of each result in "List" mode (similar to NYSL).

File: public/app/ui/components/collections/library/thumbnails-result.html [hunk 10]
```diff
@@ -0,0 +1,6 @@
+<div class="thumbnail-result thumbnail-result-library">
```
Comment (leftclickben) [inline public/app/ui/components/collections/library/thumbnails-result.html]: This is the per-collection view of each result in "Thumbnails" mode.  I don't expect there to be much variation between the collections here, if any, but doing it this way for consistency.

File: public/app/ui/components/collections/table-results.html [hunk 11]
```diff
@@ -0,0 +1,26 @@
+<table class="table table-bordered table-striped results-table">
```
Comment (leftclickben) [inline public/app/ui/components/collections/table-results.html]: And again just for consistency / cleanliness I pulled the table view out into its own component.  It's still generic to all collections, just moved out of `ui/pages/search/index.html` into its own file.

File: public/app/ui/components/controls/ListControlsComponent.js [hunk 12]
```diff
@@ -0,0 +1,157 @@
+(function () {
+    'use strict';
+
+    define(
+        [
+            'lodash',
+            'knockout',
+            'config/routes'
+        ],
+        function (_, ko, routes) {
```
Comment (leftclickben) [inline public/app/ui/components/controls/ListControlsComponent.js]: This is the UI component that controls the pager and the view mode setting.

File: public/app/ui/components/display/hierarchy-list.html [hunk 13]
```diff
@@ -1,15 +1,17 @@
-<!-- ko if: hasValue -->
-<ul>
-    <!-- ko foreach: { data: value, as: 'item' } -->
-    <li>
-        <!-- ko foreach: { data: item, as: 'hierarchyNode' } -->
-        <!-- ko if: $index() > 0 -->&rArr;<!-- /ko -->
-        <span data-bind="html: hierarchyNode"></span>
+<div class="display display-hierarchy-list">
```
Comment (leftclickben) [inline public/app/ui/components/display/hierarchy-list.html]: Adding these wrapper elements to each of these, it seemed weird having an HTML file that didn't have a single root element.

File: public/app/ui/pages/detail/index.html [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+        Loading Library record, please wait...
+    </div>
+    <!-- /ko -->
+    <!-- ko if: displayRecord -->
+    <div class="type-header" data-bind="text: typeHeader"></div>
+    <h1>
+        <span data-bind="component: displayForLabelField()"></span>
+        <span class="idno" data-bind="text: idno"></span>
+    </h1>
+    <div data-bind="component: { name: detail, params: $data }"></div>
```
Comment (leftclickben) [inline public/app/ui/pages/detail/index.html]: Includes the relevant detail view from `app/ui/components/collections/[collection]/detail.html`.

File: public/app/ui/pages/detail/index.html [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+    <!-- ko if: loading -->
+    <div class="loading text-muted">
+        <img src="/images/loading.gif" alt="" title="Loading animation" />
+        Loading Library record, please wait...
+    </div>
+    <!-- /ko -->
+    <!-- ko if: displayRecord -->
+    <div class="type-header" data-bind="text: typeHeader"></div>
+    <h1>
+        <span data-bind="component: displayForLabelField()"></span>
```
Comment (leftclickben) [inline public/app/ui/pages/detail/index.html]: Each collection now defines, through its `settings` object in the DI container, the name of the field that is considered the "label field" for use in cases like this.  (i.e. `Title` or `ItemName`).

File: public/app/ui/pages/search/index.html [hunk 16]
```diff
@@ -46,31 +46,37 @@ <h1 data-bind="text: heading"></h1>
...
-                    <td><a data-bind="attr: { href: $parent.detailUrlFor(result) }">View</a></td>
-                </tr>
-                <!-- /ko -->
-                </tbody>
-            </table>
+        <div class="col-md-12 text-right">
+            <div data-bind="component: { name: 'controls/list', params: { pager: pager, resultsMode: resultsMode, availableResultsModes: availableResultsModes, searchUrlFor: searchUrlFor } }"></div>
         </div>
     </div>
+    <div data-bind="attr: { class: resultsModeContainerClassName }">
```
Comment (leftclickben) [inline public/app/ui/pages/search/index.html]: So here's the generic parts of the three different views.

File: public/app/util/bind.js [hunk 17]
```diff
@@ -0,0 +1,48 @@
+(function () {
+    'use strict';
+
+    define(
+        [
+            'lodash',
+            'knockout'
+        ],
+        function (_, ko) {
```
Comment (leftclickben) [inline public/app/util/bind.js]: This module is aptly named.  Code is too gluey to test.

File: test/spec/config/collections/library/detailFields.spec.js [hunk 18]
```diff
@@ -15,80 +16,80 @@
                     expect(detailFields).to.have.length(17);
                 });
                 it('Defines the `idno` field', function () {
-                    expect(detailFields[0].key).to.equal('idno');
-                    expect(detailFields[0].labelText).to.equal('Accession Number');
+                    var idnoField = _.find(detailFields, { key: 'idno' });
```
Comment (leftclickben) [inline test/spec/config/collections/library/detailFields.spec.js]: This is a much better pattern, it doesn't rely on order.

**Comment by leftclickben** [comment]
Note this doesn't include sorting, but it is a complete PR.

**Comment by leftclickben** [comment]
Sorry I just noticed a problem with this, which was introduced in one of the later commits.  It's reloading the data on every pagination change, which isn't supposed to be the case.

File: public/app/ui/components/collections/library/list-result.html [hunk 19]
```diff
@@ -0,0 +1,4 @@
...
+    <span class="pull-left" data-bind="component: displayFor('MediaThumbnail', result)"></span>
+    <a href="#" data-bind="attr: { href: detailUrlFor(result) }, component: displayForLabelField(result)"></a>
```
Comment (kehh) [inline public/app/ui/components/collections/library/list-result.html]: We also need the following fields below the title:
```
^Author <br/>
^Publisher: ^PlaceOfPublication, ^DateOfPublication.
```
I'd also argue the ^PublicationType should also be included.  Maybe presented like a tag and on the right, but @mostro20 is probably best placed to make that call.

**Comment by kehh** [comment]
Also the details view doesn't load when clicking one of the links, it only loads when refreshing the page :( anyway I'll leave this to you while it's WiP.

**Comment by kehh** [comment]
Or maybe 93e5fc6 fixed that.

**Comment by leftclickben** [comment]
OK this is rooly trooly ready to go now :-)

File: public/app/ui/components/collections/library/list-result.html (hunk 19 above)
Comment (leftclickben) [inline public/app/ui/components/collections/library/list-result.html]: Yeah these views are only placeholders for @mostro20 to fill in the actual fields and layout he wants.  The details views have all the fields listed but need layout.

**Comment by leftclickben** [comment]
And for the record, yes that commit fixed that issue :-)

File: test/spec/config/collections/library/detailFields.spec.js (hunk 18 above)
Comment (kehh) [inline test/spec/config/collections/library/detailFields.spec.js]: ah so this is why you added the find by key method.

**Comment by kehh** [comment]
ouch

**Pull request merged**